libp2p/transports/tcptransport.nim

```diff
@@ -12,262 +12,325 @@
-      let useExc {.used.} = exc
-      debug "Error cleaning up client", errMsg = exc.msg, conn
+    trace "Cleaned up client", addrs = $client.remoteAddress, conn
 
   self.clients[dir].add(client)
+
   asyncSpawn onClose()
 
   return conn
 
 proc new*(
-  T: typedesc[TcpTransport],
-  flags: set[ServerFlags] = {},
-  upgrade: Upgrade,
-  connectionsTimeout = 10.minutes): T {.public.} =
-
-  let
-    transport = T(
-      flags: flags,
-      clientFlags:
-        if ServerFlags.TcpNoDelay in flags:
-          compilesOr:
-            {SocketFlags.TcpNoDelay}
-          do:
-            doAssert(false)
-            default(set[SocketFlags])
-        else:
-          default(set[SocketFlags]),
-      upgrader: upgrade,
-      networkReachability: NetworkReachability.Unknown,
-      connectionsTimeout: connectionsTimeout)
-
-  return transport
-
-method start*(
-  self: TcpTransport,
-  addrs: seq[MultiAddress]) {.async.} =
-  ## listen on the transport
-  ##
-
-  if self.running:
-    warn "TCP transport already running"
-    return
-
-  await procCall Transport(self).start(addrs)
-  trace "Starting TCP transport"
-  trackCounter(TcpTransportTrackerName)
+    T: typedesc[TcpTransport],
+    flags: set[ServerFlags] = {},
+    upgrade: Upgrade,
+    connectionsTimeout = 10.minutes,
+): T {.public.} =
+  T(
+    flags: flags,
+    clientFlags:
+      if ServerFlags.TcpNoDelay in flags:
+        {SocketFlags.TcpNoDelay}
+      else:
+        default(set[SocketFlags])
+    ,
+    upgrader: upgrade,
+    networkReachability: NetworkReachability.Unknown,
+    connectionsTimeout: connectionsTimeout,
+  )
+
+method start*(self: TcpTransport, addrs: seq[MultiAddress]): Future[void] =
+  ## Start transport listening to the given addresses - for dial-only transports,
+  ## start with an empty list
+
+  # TODO remove `impl` indirection throughout when `raises` is added to base
+
+  proc impl(
+      self: TcpTransport, addrs: seq[MultiAddress]
+  ): Future[void] {.async: (raises: [transport.TransportError, CancelledError]).} =
+    if self.running:
+      warn "TCP transport already running"
+      return
 
-  for i, ma in addrs:
-    if not self.handles(ma):
-      trace "Invalid address detected, skipping!", address = ma
-      continue
+    trace "Starting TCP transport"
 
     self.flags.incl(ServerFlags.ReusePort)
-    let server = createStreamServer(
-      ma = ma,
-      flags = self.flags,
-      udata = self)
 
-    # always get the resolved address in case we're bound to 0.0.0.0:0
-    self.addrs[i] = MultiAddress.init(
-      server.sock.getLocalAddress()
-    ).tryGet()
+    var supported: seq[MultiAddress]
+    var initialized = false
+    try:
+      for i, ma in addrs:
+        if not self.handles(ma):
+          trace "Invalid address detected, skipping!", address = ma
+          continue
 
-    self.servers &= server
+        let
+          ta = initTAddress(ma).expect("valid address per handles check above")
```

Comment:
If I understand correctly, this is going to raise a `Defect`. What's the motivation for that? I understood the objective was to eliminate all the potential sources of Defects unless it's explicitly clear we want to do that due to an unrecoverable error. This doesn't seem to be the case.